### PR TITLE
PUBDEV-5590 - Argument to limit number of sessions for import_sql_table

### PR DESCRIPTION
**Pull request by Pscheidl**
https://0xdata.atlassian.net/browse/PUBDEV-5590

### CODE REVIEW

**Comment by Pscheidl** [comment]
The idea is to limit number of SQL connections pre-allocated, as the number of sessions is derived from it. And of course, having larger local pool of connections than number of all possible sessions would be waste of resources.

Also, please note there is a minimal amout of 1 connection per node and maximal amount of 100 connections per node that is historically there and is **not** trespassed in this PR.

**Comment by Pscheidl** [comment]
@michalkurka Please have a look if my style of property naming is correct in this case.

**Comment by Pscheidl** [comment]
Will add tests.

**Comment by Pscheidl** [comment]
Tests added, made two modifier changes to make it better testable (package-private access).

Feel free to review @michalkurka

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 1]
```diff
@@ -274,6 +272,45 @@ protected void setupLocal() {
       }
     }
 
+    /**
+     * @return Number of connections to an SQL database to be opened on a single node.
+     */
+    int getMaxConnectionsPerNode() {
+      int conPerNode;
+      final String userDefinedMaxConnections = System.getProperty(MAX_USR_CONNECTIONS_KEY);
+      if (userDefinedMaxConnections != null) {
```
Comment (seb-h2o) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: useless condition: use case already handled in the try-catch below

File: h2o-core/src/test/java/water/jdbc/SQLManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  @Rule
+  public final RestoreSystemProperties restoreSystemProperties = new RestoreSystemProperties();
+
+  private Field cloudSizeField;
+  private Field currentRuntimeField;
+
+  @Before
+  public void setUp() throws Exception {
+    Paxos._commonKnowledge = true;
+    cloudSizeField = H2O.class.getField("_memary");
```
Comment (seb-h2o) [inline h2o-core/src/test/java/water/jdbc/SQLManagerTest.java]: is all this reflection setup really necessary?
after all, SqlManager uses only H2O.getCloudSize() and Runtime.getRuntime().availableProcessors().
If you're able to mock the latter in the tests below, this should work for the former as well, no?

File: h2o-core/src/main/java/water/jdbc/SQLManager.java (hunk 1 above)
Comment (Pscheidl) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: Great catch, thanks. Removed.

File: h2o-core/src/test/java/water/jdbc/SQLManagerTest.java (hunk 2 above)
Comment (Pscheidl) [reply h2o-core/src/test/java/water/jdbc/SQLManagerTest.java]: You are right. There was PowerMockito used already partially. Direct reflection manipulation has been removed and replaced by PowerMockito. Of course, there are still assertions if given values are mocked exactly as planned.

**Comment by Pscheidl** [comment]
@seb-h2o I've addressed your proposals. Everything was valid. Thank you.

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 3]
```diff
@@ -234,16 +233,19 @@ private static void initializeDatabaseDriver(String databaseType) {
...
   }
 
-  private static class SqlTableToH2OFrame extends MRTask<SqlTableToH2OFrame> {
+  static class SqlTableToH2OFrame extends MRTask<SqlTableToH2OFrame> {
     final String _url, _table, _user, _password, _columns;
     final int _numCol, _nChunks;
     final boolean _needFetchClause;
     final Job _job;
+    final int _cloudSize;
+    int _nthreads;
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: I would make this transient and initialize only in setupLocal not both in constructor and setupLocal

**Comment by Pscheidl** [comment]
@michalkurka  can we merge ?

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 4]
```diff
@@ -15,6 +15,7 @@
 public class SQLManager {
 
   private final static String TEMP_TABLE_NAME = "table_for_h2o_import";
+  private static final String MAX_USR_CONNECTIONS_KEY = H2O.OptArgs.SYSTEM_PROP_PREFIX + "sql.connections.max";
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: please use the same order of modifiers - i looks strange this way (I would change the others in this class)

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 5]
```diff
@@ -272,6 +269,43 @@ protected void setupLocal() {
       } catch (SQLException ex) {
         throw new RuntimeException("SQLException: " + ex.getMessage() + "\nFailed to connect to SQL database with url: " + _url);
       }
+      _cloudSize = H2O.getCloudSize();
```
[the remark this answers is not in the document]
Comment (michalkurka) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: This seems wrong indeed.

[the remark this answers is not in the document]
Comment (michalkurka) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: I would get rid of the fields and just use local variables that get passed into `getMaxConnectionsPerNode`. I don't see why we need them, they are used just in the context of the method.

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 6]
```diff
@@ -274,6 +269,42 @@ protected void setupLocal() {
...
+     */
+    int getMaxConnectionsPerNode(final int cloudSize, final short nThreads) {
+      int conPerNode;
+      final String userDefinedMaxConnections = System.getProperty(MAX_USR_CONNECTIONS_KEY);
+        try {
+          Integer connectionAmount = Integer.valueOf(userDefinedMaxConnections);
+          if (connectionAmount > 0 && connectionAmount < MAX_CONNECTIONS) {
+            conPerNode = calculateLocalConnectionCount(connectionAmount, cloudSize, nThreads);
+          } else {
+            conPerNode = calculateLocalConnectionCount(MAX_CONNECTIONS, cloudSize, nThreads);
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: Why do we not respect users setting? I would respect what the user wants and issue a warning

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 7]
```diff
@@ -192,8 +191,10 @@ SELECT ROW_NUMBER() OVER () AS RowNum_, <table>.* FROM <table>
     H2O.H2OCountedCompleter work = new H2O.H2OCountedCompleter() {
       @Override
       public void compute2() {
-        Frame fr = new SqlTableToH2OFrame(connection_url, finalTable, needFetchClause, username, password, columns, 
-                numCol, _v.nChunks(), j).doAll(columnH2OTypes, _v)
+        ArrayBlockingQueue<Connection> connectionPool = ConnectionPoolProvider.createConnectionPool(connection_url,
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: I believe this has to be created in the MR task in `setupLocal`

**Pull request merged**